File: 104.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

def depth(tree):
    if not tree:
        return 0
    return max(depth(tree.left),depth(tree.right))+1


class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        return depth(root)

'''
Or

class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if not root:
            return 0
        return max(maxDepth(root.left),maxDepth(root.right))+1

'''

File: xiong/104.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if(root == NULL) return 0;
        int mr = 1;
        vector<TreeNode*> s;
        vector<int> sl;
        s.push_back(root);
        sl.push_back(1);
        while(!s.empty()){
            TreeNode* now = s.back();
            int nowl = sl.back();
            mr = max(mr, nowl);
            s.pop_back();
            sl.pop_back();
            if(now->right != NULL){
                s.push_back(now->right);
                sl.push_back(nowl + 1);
            }
            if(now->left != NULL){
                s.push_back(now->left);
                sl.push_back(nowl + 1);
            }
        }
        return mr;
    }
};